### ISSUE TITLE
feat: expose `data` to template directly

### ISSUE THREAD

**Issue by jaredcwhite** [bridgetownrb/bridgetown#527]
Something I've thought about for some time is the verbosity of writing `resource.data.X` all the time. If you know you need to call `resource.data` in numerous places, why not just expose `data` directly? So instead of `resource.data.title`, just write `data.title`. We'd still provide `resource` of course because some methods are on the resource object directly like `relative_url`, `relations`, etc., plus 100% backwards compatibility.

(This would also make the following dynamic route example more concise and logical:)

```erb
---<%
# route: /items/:slug
r.get do
  item_id, *item_sku = r.params[:slug].split("-")
  item_sku = item_sku.join("-")

  render_with data: {
    layout: :page,
    title: "Item Page",
    item_id: item_id,
    item_sku: item_sku
  }
end
%>---

<p><strong>Item ID:</strong> <%= data.item_id %></p>

<p><strong>Item SKU:</strong> <%= data.item_sku %></p>
```

Thoughts?

**Comment by johlym** [comment]
Do you see a benefit to this versus something like assigning the data object to a variable in either Liquid (`{% assign data = site.resource.data.object %}`) or ERB flavor?

**Comment by jaredcwhite** [comment]
Oh sure, you can assign your own `data` variable in any template language…just feels like it should be baked into the framework as a pattern if it's that truly useful.

**Comment by johlym** [comment]
Would this be limited to its own data or could other generated data resources also be included? Tangentially, could the former be "assigned" or associated with another resource so that if this theoretical implementation existed, the related data is ready to go? 

I think in my head I'm picturing `.yml` files that can start with some sort of reserved association key a la Rails like:

```yml
# src/data/projects.yml
belongs_to: projects
projects:
  - name: Solved World Hunger
    duration: 439 years
...
```

and `projects.liquid` would automatically have access to ^ via `data.projects` without having to create the association via `projects = site.data.projects` and thus 

```liquid
# src/pages/projects.liquid
{% for project in data.projects %}
...
{{ project.duration }}
...
{% endfor %}
```

...or its other templating language equivalent would just work out of the box.

Now that I've typed all this out, I think I just described turning data files into pseudo-models. 

In ^ case, I'd use it. The hurdle for me would only be how to define what arbitrary data is accessible where in a way that's natural.

**Comment by jaredcwhite** [comment]
I like this brainstorming! You could be on to something here, which is a formal relationship between a "page" and its "dataset". Typically we'd assume that's via front matter, but in your example there it's two separate files/sources. I'll have to muse on this a bit to figure out if there's a good heuristic and terminology to wrap around the concept.

We do already have the ability to define relationships between resources in different collections, but maybe that's a little different.

**Comment by johlym** [comment]
> Typically we'd assume that's via front matter

Yeah, I mean one could very well simply set:
```yaml
---
...
projects: site.data.projects
...
---
```

in the front matter and be done with it, though if one wishes to have a data set associated with multiple pages, they'd be defining that association in multiple places compared to the inverse.

**Comment by jaredcwhite** [comment]
@johlym After some time has passed, I think your last April 5 comment provided some key insight into this feature. It's now implemented in #622! Check out the docs here: https://bridgetown-beta-pr-622.onrender.com/docs/datafiles#merging-site-data-into-resource-data
